python/segyio/_segyio.c

```diff
@@ -0,0 +1,839 @@
+                                "Number of traces is not consistent with file size. File may be corrupt.");
+
+        case SEGY_INVALID_FIELD:
+            if (args.field_count == 1) {
+                return PyErr_Format(PyExc_IndexError, "Field value out of range: %d", args.field_1);
+            } else {
+                int inline_field = args.field_1;
+                int crossline_field = args.field_2;
+                return PyErr_Format(PyExc_IndexError, "Invalid inline (%d) or crossline (%d) field/byte offset. "
+                        "Too large or between valid byte offsets.", inline_field, crossline_field);
+            }
+        case SEGY_INVALID_OFFSETS:
+            return PyErr_Format(PyExc_RuntimeError, "Found more offsets than traces. File may be corrupt.");
+
+        case SEGY_INVALID_SORTING:
+            return PyErr_Format(PyExc_RuntimeError, "Unable to determine sorting. File may be corrupt.");
+
+        case SEGY_MISSING_LINE_INDEX:
+            return PyErr_Format(PyExc_KeyError, "%s number %d does not exist.", args.name, args.field_1);
+
+        default:
+            errno = args.errno_err;
+            return PyErr_SetFromErrno(PyExc_IOError);
+    }
+}
+
+static PyObject *py_handle_segy_error(int error, int errno_err) {
+    struct error_args args;
+    args.error = error;
+    args.errno_err = errno_err;
+    args.field_1 = 0;
+    args.field_2 = 0;
+    args.field_count = 0;
+    args.name = "";
+    return py_handle_segy_error_(args);
+}
+
+static PyObject *py_handle_segy_error_with_fields(int error, int errno_err, int field_1, int field_2, int field_count) {
+    struct error_args args;
+    args.error = error;
+    args.errno_err = errno_err;
+    args.field_1 = field_1;
+    args.field_2 = field_2;
+    args.field_count = field_count;
+    args.name = "";
+    return py_handle_segy_error_(args);
+}
+
+static PyObject *py_handle_segy_error_with_index_and_name(int error, int errno_err, int index, const char *name) {
+    struct error_args args;
+    args.error = error;
+    args.errno_err = errno_err;
+    args.field_1 = index;
+    args.field_2 = 0;
+    args.field_count = 1;
+    args.name = name;
+    return py_handle_segy_error_(args);
+}
+
+// ------------ Text Header -------------
+
+static PyObject *py_textheader_size(PyObject *self) {
+    return Py_BuildValue("i", segy_textheader_size());
+}
+
+static PyObject *py_read_texthdr(PyObject *self, PyObject *args) {
+    errno = 0;
+    PyObject *file_capsule = NULL;
+    int index;
+
+    PyArg_ParseTuple(args, "Oi", &file_capsule, &index);
+
+    FILE *p_FILE = get_FILE_pointer_from_capsule(file_capsule);
+
+    char *buffer = malloc(sizeof(char) * (segy_textheader_size()));
+
+    int error = segy_read_textheader(p_FILE, buffer);
+
+    if (error != 0) {
+        free(buffer);
+        return PyErr_Format(PyExc_Exception, "Could not read text header: %s", strerror(errno));
+    }
+
+    return Py_BuildValue("s", buffer);
```

Comment:
>  When memory buffers are passed as parameters to supply data to build objects, as for the "s" and "s#" formats, the required data is copied. Buffers provided by the caller are never referenced by the objects created by Py_BuildValue(). In other words, if your code invokes malloc() and passes the allocated memory to Py_BuildValue(), your code is responsible for calling free() for that memory once Py_BuildValue() returns. 

https://docs.python.org/2.0/ext/buildValue.html

Which means that you must either free the buffer after building the value, or use a stack buffer. This function leaks.